actors/miner/src/lib.rs

```diff
@@ -1950,6 +1958,306 @@ impl Actor {
+        };
+
+        let initial_pledge = initial_pledge_for_power(
+            &base_sector_power,
+            &pledge_inputs.network_baseline,
+            &pledge_inputs.epoch_reward,
+            &pledge_inputs.network_qap,
+            &pledge_inputs.circulating_supply,
+        );
+
+        let day_reward = expected_reward_for_power(
+            &pledge_inputs.epoch_reward,
+            &pledge_inputs.network_qap,
+            &base_sector_power,
+            fil_actors_runtime::EPOCHS_IN_DAY,
+        );
+
+        let storage_pledge = expected_reward_for_power(
+            &pledge_inputs.epoch_reward,
+            &pledge_inputs.network_qap,
+            &base_sector_power,
+            INITIAL_PLEDGE_PROJECTION_PERIOD,
+        );
+
+        let mut sectors_to_add = params
+            .sectors
+            .iter()
+            .map(|sector| SectorOnChainInfo {
+                sector_number: sector.sector_number,
+                seal_proof: params.seal_proof_type,
+                sealed_cid: sector.sealed_cid,
+                deprecated_deal_ids: vec![],
+                expiration: sector.expiration,
+                activation: activation_epoch,
+                deal_weight: DealWeight::zero(),
+                verified_deal_weight: DealWeight::zero(),
+                initial_pledge: initial_pledge.clone(),
+                expected_day_reward: day_reward.clone(),
+                expected_storage_pledge: storage_pledge.clone(),
+                power_base_epoch: activation_epoch,
+                replaced_day_reward: TokenAmount::zero(),
+                sector_key_cid: None,
+                flags: SectorOnChainInfoFlags::SIMPLE_QA_POWER,
+            })
+            .collect::<Vec<SectorOnChainInfo>>();
+        let sectors_len = sectors_to_add.len();
+
+        let total_pledge = BigInt::from(sectors_len) * initial_pledge;
+        let mut needs_cron = false;
+
+        rt.transaction(|state: &mut State, rt| {
+            let current_balance = rt.current_balance();
+            let available_balance = state
+                .get_unlocked_balance(&current_balance)
+                .with_context_code(ExitCode::USR_ILLEGAL_STATE, || {
+                    "failed to calculate unlocked balance"
+                })?;
+            if available_balance < total_pledge {
+                return Err(actor_error!(
+                    insufficient_funds,
+                    "insufficient funds for aggregate initial pledge requirement {}, available: {}",
+                    total_pledge,
+                    available_balance
+                ));
+            }
+            needs_cron = !state.deadline_cron_active;
+            state.deadline_cron_active = true;
+
+            state
+                .put_sectors(store, sectors_to_add.clone())
+                .with_context_code(ExitCode::USR_ILLEGAL_STATE, || "failed to put new sectors")?;
+
+            sectors_to_add.sort_by_key(|info| info.sector_number);
```

Comment:
this is also done in `assign_sectors_to_deadline`, can we do it in just one place?